# Домашнее задание «Библиотека Pandas»

## Задание 1

Скачайте с сайта https://grouplens.org/datasets/movielens/ датасет любого размера. Определите какому фильму было выставлено больше всего оценок 5.0.

In [1]:
import pandas as pd

In [27]:
ratings = pd.read_csv('ml-latest-small/ratings.csv')
movies = pd.read_csv('ml-latest-small/movies.csv')
joined = ratings.merge(movies, on='movieId', how='left') # Объединяем файлы в одну таблицу, чтобы добавить названия фильмов
joined.head()

userId  movieId  rating  timestamp                        title  \
0       1        1     4.0  964982703             Toy Story (1995)   
1       1        3     4.0  964981247      Grumpier Old Men (1995)   
2       1        6     4.0  964982224                  Heat (1995)   
3       1       47     5.0  964983815  Seven (a.k.a. Se7en) (1995)   
4       1       50     5.0  964982931   Usual Suspects, The (1995)   

                                        genres  
0  Adventure|Animation|Children|Comedy|Fantasy  
1                               Comedy|Romance  
2                        Action|Crime|Thriller  
3                             Mystery|Thriller  
4                       Crime|Mystery|Thriller

In [33]:
rating5 = joined[ joined.rating == 5.0] # Оставляем только фильмы с рейтингом 5.0
rating5['title'].value_counts().head(1) # Считаем сколько раз встречается каждый фильм. На первой строчке фильм с наибольшим показателем


Shawshank Redemption, The (1994)    153
Name: title, dtype: int64

для датасет 'ml-25m' выдает фильм Shawshank Redemption, The (1994) с показателем 39553
для датасет 'ml-latest-small' выдает фильм Shawshank Redemption, The (1994) с показателем 153

## Задание 2

По данным файла power.csv посчитайте суммарное потребление стран Прибалтики (Латвия, Литва и Эстония) категорий 4, 12 и 21 за период с 2005 по 2010 года. Не учитывайте в расчетах отрицательные значения quantity.

In [35]:
data = pd.read_csv('power.csv')
data.head()

country  year  quantity  category
0  Austria  1996       5.0         1
1  Austria  1995      17.0         1
2  Belgium  2014       0.0         1
3  Belgium  2013       0.0         1
4  Belgium  2012      35.0         1

In [53]:
filtered_countries = data[ 
    ((data['country']=='Lithuania') | (data['country']=='Latvia') | (data['country']=='Estonia')) &
    ((data['category']==4) | (data['category']==12) | (data['category']==21)) &
    (data['year'] > 2005) & (data['year'] < 2010) & (data['quantity'] >= 0)]
print(f"Суммарное потребление стран Прибалтики: {filtered_countries['quantity'].sum()}")

Суммарное потребление стран Прибалтики: 164555.0


In [58]:
# второй вариант
data.loc[ 
    ((data['country']=='Lithuania') | (data['country']=='Latvia') | (data['country']=='Estonia')) &
    ((data['category']==4) | (data['category']==12) | (data['category']==21)) &
    (data['year'] > 2005) & (data['year'] < 2010) & (data['quantity'] >= 0)]['quantity'].sum()

164555.0

## Задание 3

Выберите страницу любого сайта с табличными данными. Импортируйте таблицы в pandas dataframe.
Примеры страниц (необязательно брать именно эти):

https://fortrader.org/quotes

https://www.finanz.ru/valyuty/v-realnom-vremeni

In [64]:
# Не совсем понял задание. Нужно просто одну команду написать?
tables = pd.read_html('https://fortrader.org/quotes')
tables[0]

Название   Цена  Изменение  Изменение(%)  Обновлено
0  EURRUB Курс евро к рублю сегодня  88.97       1.09          1.24        NaN
1       USDRUB Курс рубля к доллару  75.03       0.45          0.60        NaN

In [65]:
tables[1]

Название    Цена  Изменение  Изменение(%)  \
0       Нефть Brent Курс нефти сегодня   73.07      -0.27         -0.37   
1          Нефть WTI Цены на нефть WTI   71.65      -0.45         -0.62   
2  Природный газ Цена на природный газ    3.61       0.04          0.98   
3                                 Кофе  150.05      -0.55         -0.37   
4                             Кукуруза  539.60      -4.00         -0.74   
5                                  Соя  357.90      -2.00         -0.56   
6                               Хлопок   87.00      -0.63         -0.72   

   Обновлено  
0        NaN  
1        NaN  
2        NaN  
3        NaN  
4        NaN  
5        NaN  
6        NaN

In [66]:
tables[2]

Название   Цена  Изменение  Изменение(%)  Обновлено
0       Золото Цена золота на бирже  1 807       2.00          0.00        NaN
1  Серебро Цена на серебро на бирже  25.98      -0.16         -0.59        NaN
2         Палладий Цена на Палладий  2 811     -40.00         -1.00        NaN
3           Платина Цены на платину  1 065     -18.00         -2.00        NaN
4           Медь Цена меди на бирже   4.26      -0.06         -1.28        NaN

In [67]:
tables[3]

Название      Цена  Изменение  \
0                                  ВТБ Акции ВТБ  0.048570  -0.000830   
1                         Газпром Акции газпрома    291.00      -6.39   
2                            Лукойл Акции Лукойл   6 607.5     -174.5   
3                        Роснефть Акции Роснефти    576.50      -3.75   
4                       Сбербанк Акции Сбербанка    301.16      -2.90   
5           Сургутнефтегаз Акции Сургутнефтегаза    35.580     -1.020   
6                      Норникель Акции Норникель    24 472       -186   
7                       Аэрофлот Акции Аерофлота     67.72      -0.62   
8                    Северсталь Акции Северсталь  1 553.80     -36.20   
9                    Транснефть Акции Транснефть   172 200     -2 200   
10             Ростелеком Курс акций ростелекома    102.70      +0.66   
11                      РУСАЛ Акции РУСАЛ (RUAL)    48.850     -0.755   
12      Детский мир iDreamSky Technology Limited    142.04      -8.22   
13             Петропавловск Акции Петропавловск     22.88      -0.59   
14                    Алроса Акции АЛРОСА (ALRS)   126.880     +2.280   
15                    Магнит Акции Магнит (MGNT)   5 338.5      +28.5   
16                        НМЛК Акции НЛМК (NLMK)   224.640     -4.060   
17                  М.видео Акции М.видео (MVID)    627.20      -7.20   
18    Газпром нефть Акции "Газпром нефть" (SIBN)    440.00      -8.85   
19                        НМТП Акции НМТП (NMTP)    6.9200    -0.0500   
20            Полиметалл Акции Полиметалл (POLY)  1 636.60      -1.70   
21        Полюс Золото Акции Полюс Золото (PLZL)  14 687.5     +162.0   
22                  НОВАТЭК Акции НОВАТЭК (NVTK)  1 622.40     -17.60   
23                РусГидро Акции РусГидро (HYDR)    0.8204    -0.0458   
24                Татнефть Акции Татнефть (TATN)    516.50     -19.90   
25        АФК Система Акции АФК "Система" (AFKS)    31.091     -0.579   
26                                 МТС Акции МТС    317.90      -6.45   
27                Башнефть Акции Башнефть (BANE)   1 456.5      -38.0   
28                    Яндекс Акции Яндекс (YNDX)   5 297.0      -44.4   
29  X5 Retail Group Акции X5 Retail Group (FIVE)  2 509.00      -4.00   
30                  ФСК ЕЭС Акции ФСК ЕЭС (FEES)   0.22878   -0.00148   
31                      Мечел Акции Мечел (MTLR)     74.35      +0.15   
32                МосБиржа Акции МосБиржа (MOEX)    172.17      +0.65   
33              Интер РАО Акции Интер РАО (IRAO)      4.68      -0.03   
34              Mail.ru Акции Mail.ru (MAIL, ДР)  1 634.00     -15.00   
35                        Киви Акции Киви (QIWI)     764.5       -9.0   
36                  Лента Акции Лента (LNTA, ДР)    248.40      -2.80   
37            Тинькофф Акции Тинькофф (TCSG, др)  6 650.40     -30.20   
38                        Озон Акции Озон (Ozon)  4 085.50     -85.00   

    Изменение(%)  Обновлено  
0      -1.680162        NaN  
1      -2.150000        NaN  
2      -2.600000        NaN  
3      -0.650000        NaN  
4      -0.950000        NaN  
5      -2.787000        NaN  
6      -1.000000        NaN  
7      -0.910000        NaN  
8      -2.280000        NaN  
9      -1.000000        NaN  
10      0.650000        NaN  
11     -1.522000        NaN  
12     -5.470000        NaN  
13     -2.510000        NaN  
14      1.830000        NaN  
15      0.500000        NaN  
16     -1.775000        NaN  
17     -1.130000        NaN  
18     -1.970000        NaN  
19     -0.717400        NaN  
20     -0.100000        NaN  
21      1.100000        NaN  
22     -1.070000        NaN  
23     -5.287500        NaN  
24     -3.710000        NaN  
25     -1.828000        NaN  
26     -1.990000        NaN  
27     -2.500000        NaN  
28     -0.800000        NaN  
29     -0.160000        NaN  
30     -0.642750        NaN  
31      0.200000        NaN  
32      0.380000        NaN  
33     -0.640000        NaN  
34     -0.910000        NaN  
35     -1.200000        NaN  
36     -1

In [68]:
tables[4]

Название      Цена  Изменение  Изменение(%)  \
0           Google Акции Google (GOOGL)  2 492.80     -36.68         -1.45   
1                   IBM Акции IBM (IBM)    140.19       0.37          0.26   
2     Microsoft Акции Майкрософт (MSFT)    277.01      -2.92         -1.04   
3              Apple Акции Apple (AAPL)    142.17      -2.40         -1.66   
4                  Yandex Акции яндекса     70.71      -1.12         -1.56   
5                   Toyota Акции Тойоты    173.30      -2.35         -1.34   
6      Volkswagen Курс акций Volkswagen    198.96      -7.04         -3.42   
7            Amazon Акции Amazon (AMZN)  3 657.92     -38.66         -1.05   
8           Boeing Акции Boeing Company    229.66      -2.12         -0.91   
9              Intel Акции Intel (INTC)     55.30      -0.66         -1.18   
10            Coca-Cola Акции Coca-Cola     53.88      -0.44         -0.81   
11         Netflix Акции Netflix (NFLX)    533.58      -2.38         -0.44   
12            Pfizer Акции Pfizer (PFE)     39.15      -0.20         -0.51   
13                        AMD Акции AMD     89.70      -0.84         -0.93   
14  Walt Disney Акции Walt Disney (DIS)    171.59      -1.23         -0.71   
15           NVIDIA Акции NVIDIA (NVDA)    796.19     -18.68         -2.29   
16           PayPal Акции PayPal (PYPL)    293.53      -3.60         -1.21   
17         Facebook Акции Facebook (FB)    344.20      -6.29         -1.79   
18             Tesla Акции Тесла (TSLA)    634.02     -10.63         -1.65   
19            Twiter Цена акций твитера     67.45      -1.31         -1.91   
20    Hasbro Цена на акции Hasbro, Inc.     95.87      -1.06         -1.09   
21                               Сименс    129.84      -3.84         -2.87   
22                                 Qiwi     10.20      -0.10         -0.97   
23     Mastercard Акции Mastercard (МА)    368.60      -5.32         -1.42   
24         Moderna Акции Модерна (MRNA)    230.72       8.82          3.97   
25        Макдональдс Акции Макдональдс    233.05      -1.58         -0.67   
26                                 MSCI    551.43      -4.34         -0.78   
27                             FIXPRICE      8.34      -0.29         -3.36   
28                 Zoom Акции Zoom (ZM)    387.68      -6.23         -1.58   
29     HeadHunter Акции HeadHunter (HH)     44.88      -0.24         -0.53   

    Обновлено  
0         NaN  
1         NaN  
2         NaN  
3         NaN  
4         NaN  
5         NaN  
6         NaN  
7         NaN  
8         NaN  
9         NaN  
10        NaN  
11        NaN  
12        NaN  
13        NaN  
14        NaN  
15        NaN  
16        NaN  
17        NaN  
18        NaN  
19        NaN  
20        NaN  
21        NaN  
22        NaN  
23        NaN  
24        NaN  
25        NaN  
26        NaN  
27        NaN  
28        NaN  
29        NaN

In [70]:
dividend = pd.read_html('https://www.dohod.ru/ik/analytics/dividend')
dividend

[            Акция            Сектор            Период  \
 0    Лензолото-ао  Металлы и добыча         всего 12м   
 1    Лензолото-ао  Металлы и добыча      годовые 2020   
 2     Лензолото-п  Металлы и добыча         всего 12м   
 3     Лензолото-п  Металлы и добыча      годовые 2020   
 4      Северсталь  Металлы и добыча         всего 12м   
 ..            ...               ...               ...   
 295         Русал  Металлы и добыча  I полугодие 2021   
 296      Квадра-п            Разное      годовые 2020   
 297    Мостотрест     Строительство      годовые 2021   
 298  Обувь России            Разное      годовые 2020   
 299     TCS Group   Финансы и Банки   II квартал 2021   
 
      Выплата на акцию (прогноз)  Unnamed: 4 Unnamed: 5 Доходность выплаты  \
 0                      15219.50         NaN        RUB             62.15%   
 1                      15219.50         NaN        RUB             62.15%   
 2                       3699.27         NaN        RUB             